logs/file.go

```diff
@@ -148,30 +148,42 @@ func (w *fileLogWriter) WriteMsg(when time.Time, msg string, level int) error {
 	if level > w.Level {
 		return nil
 	}
-	hd, d, h := formatTimeHeader(when)
-	msg = string(hd) + msg + "\n"
-	if w.Rotate {
-		w.RLock()
-		if w.needRotateHourly(len(msg), h) {
-			w.RUnlock()
-			w.Lock()
+
+	logFormat := "APACHE_FORMAT"
```

Comment:
If we define a new interface like
```
type LoggerFormatter interface {
    Format(when, msg, level)
}
```
It will be more scalable, but more complicated. What do you think?
Here is a similar issue #4002 
If you don't understand Chinese, just ignore the Chinese comment, and we can discuss here.

Reply:
I agree. My fix was working but not the cleanest. I'll get working on an implementation that uses a new interface for this as it will benefit anyone looking to implement their own logging format in the future

Reply:
#4073 It's a similar issue. So allowing user custom their log info format looks like very necessary.

Reply:
Alright I think I've made a bit of progress on the issue but I'm no expert so I'd like an opinion on it before I continue to make sure I'm going in the right direction.

So the user would define some sort of function themselves possibly that would be able to manipulate the fields in `AccessLogRecord`. I think these fields would be sufficient for a user to choose from.
https://github.com/astaxie/beego/blob/7a48fbb69846515d9dc3627096590f2b7b2a1bca/logs/accesslog.go#L31-L45

From this function the output would be a single string with the formattted message in whatever way the user wants it to be. Beego could access this function within the [LogAccess](https://github.com/astaxie/beego/blob/7a48fbb69846515d9dc3627096590f2b7b2a1bca/router.go#L1019) func and then pass on the formatted message to  `beeLogger.writeMsg` ([AccessLog](https://github.com/astaxie/beego/blob/7a48fbb69846515d9dc3627096590f2b7b2a1bca/logs/accesslog.go#L65) wouldn't really be needed anymore or we could move my change from `LogAccess` into it)

```go
// LogAccess logging info HTTP Access
func LogAccess(ctx *beecontext.Context, startTime *time.Time, statusCode int) {
	...
	record := &logs.AccessLogRecord{
		...
	}
	logMessage := Format(record)
	logs.AccessLog(logMessage)
}
```

This doesn't include an interface and that's mainly because I've not used them much before but I've done a good bit of reading into them today and I do think they should be used in the actual implementation

Reply:
I think what we need to do is defining an API looks like 

`WriteMsg(when time.Time, filePath string, msg string, level int) error`

The key point is `writeMsg` and `writeToLoggers` method of `BeeLogger`.

we need to change the implementation like:

```go
func (bl *BeeLogger) writeMsg(logLevel int, msg string, v ...interface{}) error {
 // ... so somthing 
file :=""

if bl.enableFuncCallDepth {
  // read file full name
}

bl.writeToLoggers(when, msg, logLevel, file)

}


func (bl *BeeLogger) writeToLoggers(when time.Time, msg string, level int, file string) {
	for _, l := range bl.outputs {
		err := l.WriteMsg(when, msg, level, file)
		if err != nil {
			fmt.Fprintf(os.Stderr, "unable to WriteMsg to adapter:%v,error:%v\n", l.name, err)
		}
	}
}
```
And then we change the interface:
```go
type Logger interface {
	Init(config string) error
	WriteMsg(when time.Time, msg string, level int, file string) error
	Destroy()
        SetFormatter(fmt Formatter)
	Flush()
}

type Formatter interface {
    Format(when time.Time, msg string, level int, file string) string
}
```
The implementation of this `Logger#WriteMsg` looks like:
```go
WriteMsg(...) {
    msg = fmtter.Format(...)
    doWrite(msg)
}
```

And we need to add new method:
```go
func SetLogger(adapter string, formatter, config ...string) error {
...
}
```

In this case, JSON log, users may invoke `SeLogger(xxx, jsonFormat, config...)`
```go
type JsonFormatter struct {
}

func (jf JsonFormatter) Format(...) string {
    if isJson(...) {
   } else {
   ...
}
}
```

Reply:
we could schdule a zoom meeting to have further discussion if you want.

Reply:
It's a big change, we could release this feature in v2.0.1 (next is v2.0.0) since we decide to refactor move log to pkg/log and then refactor logs module. If you want to be responsible for refactoring log module, we can put this in v2.0.0. 

Now we are focusing on refactoring ORM module :)

Reply:
I think a call would definitely be beneficial. It's quite a large change and I'm not too sure how long it'd take to implement fully or how much I could do myself but I'd be up for giving it my best effort. 

I'll write down some notes as to some of the questions I have and ideas to talk about during a call. I'll email you to schedule that.